From: Carlos Garnacho <carlosg@gnome.org>
Date: Thu, 25 Sep 2025 12:08:01 +0200
Subject: wayland: Check modifier state from event prior to event delivery

Refactor the code handling possible modifier state changes prior to event
delivery so they all happen in handle_event(), and use the modifier state
from incoming events there, in order to ensure the client is in sync with
the expected modifiers for the upcoming event.

After the key event is delivered, we still do check the keymap state for
the most recent modifier state (e.g. after the key(s) were handled). This
risks more wl_keyboard.modifier changes than strictly necessary, but there
should an additional guarantee about it: That a wl_keyboard.key will always
have a last wl_keyboard.modifiers matching the ClutterEvent state.

In exchange for not doing event delivery during update(), we also need to
ensure the modifiers are up to date when entering a new surface, e.g. after
alt-tab. Generally, we shouldn't be doing Wayland protocol send*() calls
from update(), so this is a move in the right direction.

Bug-Ubuntu: https://bugs.launchpad.net/bugs/2125866,
            https://bugs.launchpad.net/bugs/2125240
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4686
---
 src/wayland/meta-wayland-keyboard.c | 65 +++++++++++++++++++++++++++++--------
 1 file changed, 51 insertions(+), 14 deletions(-)

diff --git a/src/wayland/meta-wayland-keyboard.c b/src/wayland/meta-wayland-keyboard.c
index ac22adf..d7e044c 100644
--- a/src/wayland/meta-wayland-keyboard.c
+++ b/src/wayland/meta-wayland-keyboard.c
@@ -527,8 +527,28 @@ update_pressed_keys (struct wl_array *keys,
     }
 }
 
-static void
-maybe_update_modifiers (MetaWaylandKeyboard *keyboard)
+static gboolean
+maybe_update_modifiers (MetaWaylandKeyboard *keyboard,
+                        ClutterModifierType  pressed,
+                        ClutterModifierType  latched,
+                        ClutterModifierType  locked)
+{
+  if (keyboard->xkb_info.modifiers.pressed != pressed ||
+      keyboard->xkb_info.modifiers.latched != latched ||
+      keyboard->xkb_info.modifiers.locked != locked)
+    {
+      keyboard->xkb_info.modifiers.pressed = pressed;
+      keyboard->xkb_info.modifiers.latched = latched;
+      keyboard->xkb_info.modifiers.locked = locked;
+
+      return TRUE;
+    }
+
+  return FALSE;
+}
+
+static gboolean
+maybe_update_modifiers_from_keymap (MetaWaylandKeyboard *keyboard)
 {
   MetaBackend *backend = backend_from_keyboard (keyboard);
   ClutterBackend *clutter_backend = meta_backend_get_clutter_backend (backend);
@@ -541,16 +561,27 @@ maybe_update_modifiers (MetaWaylandKeyboard *keyboard)
                                      &latched,
                                      &locked);
 
-  if (keyboard->xkb_info.modifiers.pressed != pressed ||
-      keyboard->xkb_info.modifiers.latched != latched ||
-      keyboard->xkb_info.modifiers.locked != locked)
-    {
-      keyboard->xkb_info.modifiers.pressed = pressed;
-      keyboard->xkb_info.modifiers.latched = latched;
-      keyboard->xkb_info.modifiers.locked = locked;
+  return maybe_update_modifiers (keyboard, pressed, latched, locked);
+}
 
-      notify_modifiers (keyboard);
-    }
+static gboolean
+maybe_update_modifiers_from_event (MetaWaylandKeyboard *keyboard,
+                                   const ClutterEvent  *event)
+{
+  ClutterModifierType pressed, locked, latched;
+  ClutterEventType event_type;
+
+  event_type = clutter_event_type (event);
+
+  if (event_type != CLUTTER_KEY_PRESS && event_type != CLUTTER_KEY_RELEASE)
+    return FALSE;
+
+  clutter_event_get_key_state (event,
+			       &pressed,
+			       &latched,
+			       &locked);
+
+  return maybe_update_modifiers (keyboard, pressed, latched, locked);
 }
 
 void
@@ -566,8 +597,6 @@ meta_wayland_keyboard_update (MetaWaylandKeyboard   *keyboard,
   if (!update_pressed_keys (&keyboard->pressed_keys, evdev_code, is_press))
     return;
 
-  maybe_update_modifiers (keyboard);
-
   xkb_state_update_key (keyboard->xkb_info.state,
                         hardware_keycode,
                         is_press ? XKB_KEY_DOWN : XKB_KEY_UP);
@@ -599,6 +628,10 @@ meta_wayland_keyboard_handle_event (MetaWaylandKeyboard   *keyboard,
               is_press ? "press" : "release",
               hardware_keycode);
 
+  if (maybe_update_modifiers_from_event (keyboard,
+                                         (const ClutterEvent *) event))
+    notify_modifiers (keyboard);
+
   handled = meta_wayland_keyboard_broadcast_key (keyboard,
                                                  (const ClutterEvent *) event);
 
@@ -612,7 +645,8 @@ meta_wayland_keyboard_handle_event (MetaWaylandKeyboard   *keyboard,
                   "No wayland surface is focused, continuing normal operation");
     }
 
-  maybe_update_modifiers (keyboard);
+  if (maybe_update_modifiers_from_keymap (keyboard))
+    notify_modifiers (keyboard);
 
   return handled;
 }
@@ -671,6 +705,7 @@ broadcast_focus (MetaWaylandKeyboard *keyboard,
   wl_keyboard_send_enter (resource, keyboard->focus_serial,
                           keyboard->focus_surface->resource,
                           &keyboard->pressed_keys);
+
   keyboard_send_modifiers (keyboard, resource, keyboard->focus_serial);
 }
 
@@ -734,6 +769,8 @@ meta_wayland_keyboard_set_focus (MetaWaylandKeyboard *keyboard,
           keyboard->focus_serial =
             meta_wayland_input_device_next_serial (input_device);
 
+          maybe_update_modifiers_from_keymap (keyboard);
+
           wl_resource_for_each (resource, &keyboard->focus_resource_list)
             {
               broadcast_focus (keyboard, resource);
